### ISSUE TITLE
Add @select attribute in att.dataPointing class

### ISSUE THREAD

**Issue by pe-ro** [music-encoding/music-encoding#1374]
Currently to point from metadata elements to elements elsewhere in the file that they apply to, one must use the `@data` attribute.  I propose adding a `@select` attribute, also in the att.dataPointing class, which allows an XPath expression as its content.  This would make it possible to select the "pointed to" elements not just by their xml:id, but also by their position in the file hierarchy.  This would also make it possible to select a set of elements, even when none of the elements in the set have an id.  Such an attribute would reduce reliance on IDs and make MEI files smaller.

Also, the set of elements carrying the attributes from att.dataPointing can be expanded.  For instance, allowing `@select` on `<symbolDef>` would make it easy to associate an SVG shape with multiple elements, perhaps widely scattered throughout the encoding.  There may be other places where allowing att.dataPointing may be desirable.

I think `@select` will certainly be an integral part of any MEI style language that we might develop in the future.

**Comment by ahankinson** [comment]
Questions:

 - Would `@select` be limited to XPath 1? XPath 2 is not supported in browsers, for example, so the content of `@select` would need to be targeted at the lowest common execution platform...
 - Do `@select` statements assume they start from the root? Or are they relative to the location of the element that contains the `@select` statement?
 - `@plist` already captures a list of IDs, and they don't have to be of a specific type. I guess the difference is that `@select` would get executed dynamically, and `@plist` would need to be captured at encoding time? And also `@plist` relies on IDs...
 - "Such an attribute would reduce reliance on IDs and make MEI files smaller." I didn't know this was a goal? IDs make a very useful reference system. I can see how they can make MEI files more aesthetically pleasing for humans, but between large HD storage, faster networks, and efficient compression, I'm not sure the size difference would make a dramatic difference in the larger picture. Also, we'll still need IDs for doing standoff markup, and they're a much more efficient pointing system than writing a whole XPath statement to target specific elements.
 - "...integral part of any MEI style language ..." What sort of MEI-style language are you picturing?

**Comment by lpugin** [comment]
Maybe I can give some answers.

> - Would `@select` be limited to XPath 1? XPath 2 is not supported in browsers, for example, so the content of `@select` would need to be targeted at the lowest common execution platform...

This is a good point. We probably want to limit it to XPath 1.0 to start with. We can always extend to XPath 2.0, which would be not breaking backward compatibility. It would not if we change the other way around (allowing XPath 2.0 and then restricting to XPath 1.0)

> - Do `@select` statements assume they start from the root? Or are they relative to the location of the element that contains the `@select` statement?

It seems to me that it should be from the root, but you are right that this needs clarification. Having it from the point where it is specified would probably require for most of the XPath to go "up", meaning most of the needing to have some `ancestor::` parameters in them. Any problem with having it from the root?

>  - `@plist` already captures a list of IDs, and they don't have to be of a specific type. I guess the difference is that `@select` would get executed dynamically, and `@plist` would need to be captured at encoding time? And also `@plist` relies on IDs...

Yes, exactly.

>  - "Such an attribute would reduce reliance on IDs and make MEI files smaller." I didn't know this was a goal? IDs make a very useful reference system. I can see how they can make MEI files more aesthetically pleasing for humans, but between large HD storage, faster networks, and efficient compression, I'm not sure the size difference would make a dramatic difference in the larger picture. Also, we'll still need IDs for doing standoff markup, and they're a much more efficient pointing system than writing a whole XPath statement to target specific elements.

I don't think it is about making files smaller. It is more about providing an encoding approach that allow for specifying to which elements a certain characteristic (style?) applies without having to have an explicit ID for each relationship. For example, if you have a file with 1,000 notes, and want to specify that all notes with a certain octave and pitch are represented with a particular symbol, now you can either:
* make every note with the specific octave and pitch have a `@altsym` pointing to the symbol, or
* have a symbol with a `plist` with hundreds of IDs in it.

Both are perfectly fine and processable. The use of `@select` would simply provide a more global and more flexible way to do so by having the references built dynamically, as you mentioned it above. The file size seems secondary to me, and I do not see this as a proposal for a replacement of IDs. Where IDs work perfectly well (e.g., standoff markup for the start and end of a slur), I do not imagine changing to an XPath and I do not think having `@select` would force one to do so. However, there are cases where writing an XPath statement seems more efficient than writing and maintaining explicit ID references when these could very easily be expressed with XPath and resolved dynamically.

> - "...integral part of any MEI style language ..." What sort of MEI-style language are you picturing? 

I leave this for @pe-ro to answer, but I think what was meant it that if we have, one day, an MEI styling language, this approach might very well be used in it. But it does not have to, or course.

**Comment by ahankinson** [comment]
> We can always extend to XPath 2.0, which would be not breaking backward compatibility.

If a file containing an XPath 2+ statement gets read by a processor that does not have XPath 2+ capabilities, then it misses part of the encoding. If that encoding is used to, for example, select members of a slur or something "structural" then the processor has no way to actually resolve the entities in the file such that it can be understood. This breaks all kinds of compatibility.

AFAIK, there is also no way to say that something is XPath 2+ unless you try to execute it and fail.

**Comment by lpugin** [comment]
That's a good point. We should probably stick with XPath 1.0, then.

**Comment by ahankinson** [comment]
Out of interest, here is a list of XPath syntax supported in the Python standard library, which seems to not even include the full XPath 1.0 support: 

https://docs.python.org/3/library/xml.etree.elementtree.html#supported-xpath-syntax

I think this means that `ancestor::` support is not included.

**Comment by pe-ro** [comment]
> "...integral part of any MEI style language ..." What sort of MEI-style language are you picturing?
>> I leave this for @pe-ro to answer, but I think what was meant it that if we have, one day, an MEI styling language, this approach might very well be used in it. But it does not have to, or course.

I've been thinking about a stylesheet language for MEI for awhile now but it's not ready for public consumption yet.  Selection of nodes via XPath plays a significant role; that is, lots of nodes can be selected without having to list their individual IDs.

On the question of whether to support XPath 1 or 2, we can 
- decide now; or
- add an extra attribute in dataPointing or create an element within the header (probably in `<encodingDesc>` )where one can explicitly state what kind of XPath support is expected.

Poor support in Python (or any other language) is not a reason to shy away from XPath, in my opinion.

**Comment by lpugin** [comment]
FWIW, I think that the [lxml](https://lxml.de/) Python library supports XPath 1.0, so this could be a work around for people working with Python and having problems with XPath queries not supported by the standard library.

**Comment by ahankinson** [comment]
Yeah, I was specifically thinking of music21, which has support for MEI but a policy of more dependencies

**Comment by lpugin** [comment]
@pe-ro : what about having a dedicated `att.dataSelecting` class for it and only add to `symbolDef` to start with? I think the use case is pretty clear and constraint for this element, but maybe not so much for other element members of `att.dataPointing`.

**Comment by pe-ro** [comment]
That would be fine as well.

**Comment by pe-ro** [comment]
@lpugin, it's not clear to me how this proposal breaks backward compatibility? Can you explain please?

**Comment by lpugin** [comment]
I guess it depends how backward compatibility is understood. Nothing would be removed nor modified, so you are right we can remove the label.

**Comment by bwbohl** [comment]
Anyone fancy to create a PR for this and make it ready for MEI v5.1?

**Comment by lpugin** [comment]
There is [this](https://github.com/music-encoding/music-encoding/pull/1398) one.
